module/zfs/zpl_file.c

```diff
@@ -361,8 +361,10 @@ zpl_write(struct file *filp, const char __user *buf, size_t len, loff_t *ppos)
 	ssize_t wrote;
 
 	crhold(cr);
+	spl_inode_lock(file_inode(filp));
 	wrote = zpl_write_common(filp->f_mapping->host, buf, len, ppos,
 	    UIO_USERSPACE, filp->f_flags, cr);
+	spl_inode_unlock(file_inode(filp));
```

Comment:
Unlike other Linux filesystem the `zfs_range_lock()` and `zfs_range_unlock()` functions are responsible for locking the contents of a file.  For certain workloads this can be a big performance win because it's a range lock which allows different process to update different regions of the file concurrently.  Most Linux filesystems which rely on the inode lock end up serializing everything unnecessarily.  We only need to hold this lock when updating the inode, not during the write itself.